File: 0079-word-search/0079-word-search.cpp
class Solution {
public:
    string word;
    int ROWS; 
    int COLS;
    vector<vector<char>> board;
    vector<vector<int>> dir = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

    bool dfs(int r, int c, int idx) {
        // Base cases
        if (idx == word.size()) return true;
        if (r < 0 || r >= ROWS || c < 0 || c >= COLS || board[r][c] != word[idx]) return false;

        // Temporarily mark the cell as visited
        char temp = board[r][c];
        board[r][c] = '#'; // Use a special character to mark as visited

        // Explore all possible directions
        for (auto d : dir) {
            int row = r + d[0];
            int col = c + d[1];
            if (dfs(row, col, idx + 1)) return true;
        }

        // Restore the cell's original value
        board[r][c] = temp;
        return false;
    }

    bool exist(vector<vector<char>>& board, string word) {
        this->word = word;
        ROWS = board.size();
        COLS = board[0].size();
        this->board = board;

        // Start DFS from every cell that matches the first letter of the word
        for (int r = 0; r < ROWS; ++r) {
            for (int c = 0; c < COLS; ++c) {
                if (board[r][c] == word[0] && dfs(r, c, 0)) return true;
            }
        }

        return false;
    }
};

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        
        ROWS, COLS = len(board), len(board[0])
        dir = [[0,1],[0,-1],[1,0],[-1, 0]]
        dp = {}
        seen = set()
        
        def dfs(r, c, idx):
            
            if board[r][c] != word[idx]:
                return False
            
            if idx == len(word) - 1:
                return True

            for dr,dc in dir:
                row, col = r + dr, c + dc
                if 0 <= row < ROWS and 0 <= col < COLS and (row, col) not in seen:
                    seen.add((r, c))
                    if dfs(row, col, idx + 1):
                        return True
                    seen.remove((r,c))
            
            return False
        
        for r in range(ROWS):
            for c in range(COLS):
                if dfs(r,c,0):
                    return True
        
        return False

